DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,478 @@
+
+3. Introduce a preblit function that will be called before blitting the fields.
+The purpose of the preblit is to offer the possibility of preparing the data
+before the blitting phase; acquiring the mutex on the `struct` that will be
+copied is one of the operations that the preblit will be responsible for. Later
+on, the mutex may be released in the postblit. This approach has the benefit of
+minimizing the mutex protected area in a manner that offers encapsulation, but
+suffers the disadvantage of adding even more complexity on top of existing one by
+introducing a new concept which requires typecheking of disparate sections of
+code (need to typecheck across preblit, mempcy and postblit).
+
+4. Use an implicit global mutex to synchronize the blitting of fields. This approach
+has the advantage that the compiler will do all the work and synchronize all blitting
+phases (even if the threads don't actually touch each other's data) at the cost
+of performance. Python implements a global interpreter lock and it was proven
+to cause unscalable high contention; there are ongoing discussions of removing it
+from the Python implementation [5].
+
+### Introducing the copy constructor
+
+As stated above, the fundamental problem with the postblit is the automatic blitting
+of fields which makes it impossible to type check and cannot be synchronized without
+additional costs.
+
+As an alternative, this DIP proposes the implementation of a copy constructor. The benefits
+of this approach are the following:
+
+* it is a known concept. C++ has it [3];
+* it can be typechecked as a normal constructor (since no blitting is done, the data is
+initialized as if it were in a normal constructor); this means that `const`/`immutable`/`shared`
+copy constructors will be type checked exactly as their analogous constructors
+* offers encapsulation
+
+The downside of this solution is that the user must do all the field copying by hand
+and every time a field is added to a struct, the copy constructor must be modified.
+However, this can be easily bypassed by D's introspection mechanisms. For example,
+this simple code may be used as a language idiom or library function:
+
+```D
+static foreach (i, ref field; src.tupleof)
+    this.tupleof[i] = field;
+```
+
+As shown above, the single benefit of the postblit can be easily substituted with a
+few lines of code. On the other hand, to fix the limitations of the postblit it is
+required that more complexity is added for little to no benefit. In these circumstances,
+for the sake of uniformity and consistency, replacing the postblit constructor with a
+copy constructor is the reasonable thing to do.
+
+## Description
+
+This section discusses all the technical aspects regarding the semantics
+of the copy constructor.
+
+### Syntax
+
+The following link exhibits the proposed syntax and the necessary grammar changes in order to support
+the copy construct:
+
+https://github.com/dlang/dlang.org/pull/2414
```

Comment:
Why not something similar to C++? It's not even discussed here. Also, making the syntax a link to somewhere else makes it harder to read when it's an important part of the DIP.

Reply:
The link is added in order to show exactly what the changes to the grammar are and also because there is no nice way to explicitly express the grammar in the md format.

Reply:
It makes it a lot harder to read and understand what's being proposed.

I still don't understand why such a radical departure from C++ syntax, given that AFAIK is the only language with a copy constructor.